[FEATURE] Contributing a Step-by-Step Guide Using Your Browser and GitHub (#15)
TYPO3-Documentation/TYPO3CMS-Guide-StepByStep: Documentation/90Contribute/20ContributorGuide/35ContributingAStepByStepGuideUsingYourBrowserAndGitHub.md

# Contributing a Step-by-Step Guide Using Your Browser and GitHub

<!-- #Beginner @mathias -->

It's easy to add a new guide to the TYPO3 Step-by-Step Guides. All of the guides are written in plain text using simple formatting instructions called Markdown. You can add a new guide using only your web browser and the tools available on the GitHub website, where the guide files are stored.

## Learning objective

In this step-by-step guide you will create a new Step-by-Step guide and create a *pull request*, asking for your guide to be included in the TYPO3 Step-by-Step guides.

## Prerequisites

### Tools and technology

* A computer with a web browser and an internet connection

### Knowledge and skills

* How to use a web browser
* How to edit files and make screenshots on the computer you are using
* Basic understanding of Markdown, as explained in the [Markdown Cheat Sheet](https://docs.typo3.org/permalink/h2document:md-cheat-sheet)
* How to use the step-by-step guide template, as explained in [Using the Template](30UsingTheTemplate.md).

> ## Create a GitHub user

Reviewer: ```suggestion
## Create a GitHub user account
```

> [!NOTE]
> You can skip this section if you already have a user account on GitHub.
> Continue to the *Create a new file* section below.

The TYPO3 Step-by-Step guide files are stored online, at a website called GitHub. This is a place where people upload and collaborate on collections of files and folders relating to their file-based projects.

To edit files at GitHub, you need to create a user account there. Let's create one.

1. In your browser, visit [Creating an account on GitHub](https://docs.github.com/en/get-started/start-your-journey/creating-an-account-on-github) in the GitHub Documentation.
2. Read the information on the page.
3. Follow the 4 steps in the *Signing up for a new personal account* section.

You now have a user account on GitHub. At the end of the sign-up process, you should be logged in as a user. You must always be logged in on GitHub in order to edit and contribute to files in repositories and projects there.

## Copy content from the template

> All step-by-step guides should be based on the template file. It includes important instructions and makes the job of writing the guide easier. You need to copy it so it can be the strating point for your new guide.

Reviewer: ```suggestion
All step-by-step guides should be based on the template file. It includes important instructions and makes the job of writing the guide easier. You need to copy it so it can be the starting point for your new guide.
```

1. Visit the step-by-step guide template in your web browser by clicking this link: [github.com/TYPO3-Documentation/TYPO3CMS-Guide-StepByStep/blob/contrib/Documentation/90Contribute/10Template/Index.md](https://github.com/TYPO3-Documentation/TYPO3CMS-Guide-StepByStep/blob/contrib/Documentation/90Contribute/10Template/Index.md)

> 2. Whent he page has loaded, click the "Copy raw file" button. The button has no text, but an icon with two partially overlapping squares. In the screenshot below, it is highlighted with an orange frame and there is an orange arrow pointing to it.

Reviewer: ```suggestion
2. When the page has loaded, click the "Copy raw file" button. The button has no text, but an icon with two partially overlapping squares. In the screenshot below, it is highlighted with an orange frame and there is an orange arrow pointing to it.
```

> [!NOTE]
> Don't worry, if GitHub displays a message like this one. This is just as expected.
> ![Blue box with the text "You’re making changes in a project you don’t have write access to. Submitting a change will write it to a new branch in your fork my-username/TYPO3CMS-Guide-StepByStep, so you can send a pull request."](Images/45ContributingAStepByStepGuideUsingYourBrowserAndGitHub/GitHubYouDontHaveAccess.png)

2. Where it says "Enter file contents here" is a text editor. Paste in the step-by-step guide template that you copied previously.
3. Enter a name for your file in the text box where it says "Name your file…". The [Using the Template](30UsingTheTemplate.md) guide explains exactly what your file name should be, how it should be formatted, and that it must end in ".md".

When you are finished, it should look something like in this screenshot:

![Screenshot of a GitHub interface showing a Markdown file titled "CreatingAGuide.md" being edited. The file contains a template for a TYPO3 step-by-step guide, with instructional text and placeholders.](Images/45ContributingAStepByStepGuideUsingYourBrowserAndGitHub/GitHubNewGuideFileWithContent.png)

> 

Reviewer: ```suggestion
## Write the guide

Here is the point in the process when you write the step-by-step guide. See [Using the Template] 30UsingTheTemplate.md) for help to write each section of the guide. 

```

## Commit your changes

> Committing your changes means the same as saving them, but you also add a message saying what changes your save action includes.

Reviewer: ```suggestion
Committing your changes means the same as saving them, but you also add a message saying what changes your save action includes.

> [!TIP]
> You don't have to wait until you've finished writing the guide to do one big commit. You can do lots of small commits to save as you go. You are working in your fork of the repository, so nobody knows about your work yet but you. 
```

I feel like we might need to say something like this, because I think otherwise we're implying that the person keeps the browser tab window open until they're done and what if they accidentally close the tab (lose changes) or have a power outage (lose changes).

![A dialog box titled Propose changes shows fields for a commit message, extended description, and commit email, with Cancel and green Propose changes buttons at the bottom.](Images/45ContributingAStepByStepGuideUsingYourBrowserAndGitHub/GitHubCommitMessageDialog.png)

2. Fill in the fields in the dialog:
   * **Commit message:** Briefly describe what you've done. Use just a few words or the title of your Step-by-Step guide.
   * **Extended description:** Optional. You can add a longer description here.
   * **Commit Email:** You can usually leave this unchanged. This is the email address that will be related to your commit.
3. Click the green "Propose changes" button. This will open a new page titled "Comparing changes," looking something like the screenshot below:

![A GitHub interface shows the Comparing changes view, allowing selection of repositories and branches. Options to create a pull request or compare forks are visible, along with commit, file changed, and contributor details.](Images/45ContributingAStepByStepGuideUsingYourBrowserAndGitHub/GitHubComparingChanges.png)

## Open a pull request

> To tell the Step-by-Step Guide maintainers that you have new things to contribute, you must open a *pull request*. Basically a request to "please, pull in my changes."

Reviewer: ```suggestion
When you're finished writing your guide, you're ready to ask for it to be copied from your fork into the official step-by-step guide repository. 

To tell the Step-by-Step Guide maintainers that you have new things to contribute, you must open a *pull request*. Basically a request to "please, pull in my changes."
```

> You're still on the "Comparing changes" page shown in the screenshot above.

Reviewer: We might need to slightly re-word this if we accept my suggestions about doing multiple commits along the way.

Author: Acknowledged. I'm keeping this in mind for a future modification.
